### PR TITLE
check active annotation

### PR DESCRIPTION
**Pull request by droslean**
Adds the logic to use the `ci.openshift.io/active` annotation if it's after the latest pod transition time.

- [X] Update README
- [X] Add test cases.

### CODE REVIEW

File: pkg/controller/ttl_manager.go [hunk 1]
```diff
@@ -320,6 +321,18 @@ func resolveTtlStatus(ns *coreapi.Namespace, processPods func() (bool, time.Time
 		status.active = active
 		status.logger = status.logger.WithField("active", active)
 		if !lastTransitionTime.IsZero() {
+			// If the namespace has an `active` annotation and
+			// it's after the last transition time, use that instead.
+			if activeTimeString, exists := ns.ObjectMeta.Annotations[activeAnnotation]; exists {
+				activeTime, err := time.Parse(time.RFC3339, activeTimeString)
+				// in case of parsing error, the last transition time will be used anyways.
+				if err == nil {
```
Comment (stevekuznetsov) [inline pkg/controller/ttl_manager.go]: nit: idiom would be

```go
if activeTime, err := time.Parse(time.RFC3339, activeTimeString); err == nil {
```

File: pkg/controller/ttl_manager.go [hunk 2]
```diff
@@ -320,6 +321,18 @@ func resolveTtlStatus(ns *coreapi.Namespace, processPods func() (bool, time.Time
 		status.active = active
 		status.logger = status.logger.WithField("active", active)
 		if !lastTransitionTime.IsZero() {
+			// If the namespace has an `active` annotation and
+			// it's after the last transition time, use that instead.
+			if activeTimeString, exists := ns.ObjectMeta.Annotations[activeAnnotation]; exists {
+				activeTime, err := time.Parse(time.RFC3339, activeTimeString)
+				// in case of parsing error, the last transition time will be used anyways.
```
Comment (stevekuznetsov) [inline pkg/controller/ttl_manager.go]: Explain this best-effort approach in the README

File: pkg/controller/ttl_manager.go [hunk 3]
```diff
@@ -320,6 +321,18 @@ func resolveTtlStatus(ns *coreapi.Namespace, processPods func() (bool, time.Time
...
+				if err == nil {
+					status.logger = status.logger.WithField("active-at", activeTime.Format(time.RFC3339))
+					if activeTime.After(lastTransitionTime) {
+						lastTransitionTime = activeTime
```
Comment (stevekuznetsov) [inline pkg/controller/ttl_manager.go]: We have a number of cases to test for `resolveTtlStatus` now:

 - active annotation ignored when namespace is active
 - active annotation parsing error is ignored
 - active annotation is ignored when it's before the last transition time
 - active annotation is used when it's after the last transition time

File: pkg/controller/ttl_manager.go (hunk 1 above)
Comment (stevekuznetsov) [reply pkg/controller/ttl_manager.go]: also might want to log a warning here

**Review by stevekuznetsov — CHANGES_REQUESTED** [review]

File: pkg/controller/ttl_manager.go [hunk 4]
```diff
@@ -320,6 +321,18 @@ func resolveTtlStatus(ns *coreapi.Namespace, processPods func() (bool, time.Time
 		status.active = active
 		status.logger = status.logger.WithField("active", active)
 		if !lastTransitionTime.IsZero() {
```
Comment (petr-muller) [inline pkg/controller/ttl_manager.go:323]: I don't have enough knowledge of the code, but is it possible for `lastTransitionTime` to be zero while active annotation is present? Would we still want use the active annotation + softttl in that case?

Comment (stevekuznetsov) [reply pkg/controller/ttl_manager.go:323]: no, because soft TTL only applies when the namespace is not active (when last transition time is nonzero)

**Comment by stevekuznetsov** [comment]
Forgot to get the bots access to this one ...
/close

File: README.md [hunk 5]
```diff
@@ -14,6 +14,9 @@ Whichever TTL is reached first will be used by the controller to delete the name
 The controller will will only consider pods matching the `!ci.openshift.io/ttl.ignore` selector for soft TTL calculations, allowing
 non-matching pods to keep running while considering the namespace to be inactive.
 
+In additional, if the `ci.openshift.io/active` annotation exists and it's after the latest pod's transition time, the controller
```
Comment (stevekuznetsov) [inline README.md]: Maybe something more like the following, in the description in the table:

```
The time after `Pod` termination or last active marker where, if no other `Pod`s are `Pending` or `Running`, the namespace should be reaped. The last active marker is optional and stored as a RFC3339-formatted date in the `ci.openshift.io/active` annotation.
```

File: pkg/controller/ttl_manager.go (hunk 3 above)
Comment (stevekuznetsov) [reply pkg/controller/ttl_manager.go]: Still missing

 - active annotation ignored when namespace is active

File: README.md [hunk 6]
```diff
@@ -14,6 +14,8 @@ Whichever TTL is reached first will be used by the controller to delete the name
 The controller will will only consider pods matching the `!ci.openshift.io/ttl.ignore` selector for soft TTL calculations, allowing
 non-matching pods to keep running while considering the namespace to be inactive.
 
+In addition, the time after `Pod` termination or last active marker where, if no other `Pod`s are `Pending` or `Running`, the namespace should be reaped. The last active marker is optional and stored as a RFC3339-formatted date in the `ci.openshift.io/active` annotation.
```
Comment (stevekuznetsov) [inline README.md]: This should be the description column for soft ttl in the table above

File: pkg/controller/ttl_manager_test.go [hunk 7]
```diff
@@ -203,6 +203,86 @@ func TestResolveTtlStatus(t *testing.T) {
 			expectedErr:   true,
 			expectedRetry: true,
 		},
+		{
+			name: "last transition before the activeAnnotation",
```
Comment (stevekuznetsov) [inline pkg/controller/ttl_manager_test.go]: please say why _and_ what in these names -- see my tests above. this one could be:

> last transition is before the activeAnnotation *so it is ignored*

**Review by stevekuznetsov — COMMENTED** [review]
Mostly LGTM, small nits.

@petr-muller do you want to review?

File: README.md (hunk 6 above)
Comment (droslean) [reply README.md]: oops, I didn't notice it.

File: README.md [hunk 8]
```diff
@@ -8,12 +8,15 @@ the namespace either at creation time or at any later point:
 | -------- | -------------------------- | -------------------------------- | ---------------------------------------------------------------------------------------------------------------------- |
 | Hard     | `ci.openshift.io/ttl.hard` | `time.Duration`-formatted string | The time after namespace creation when the namespace should be reaped.                                                 |
 | Soft     | `ci.openshift.io/ttl.soft` | `time.Duration`-formatted string | The time after `Pod` termination where, if no other `Pod`s are `Pending` or `Running`, the namespace should be reaped. |
+| Active   | `ci.openshift.io/active`   | `time.Time`	 -formatted string | The last active marker is optional and stored as a RFC3339-formatted date.												|
```
Comment (stevekuznetsov) [inline README.md]: The table shows you TTL types... this is not a TTL type. I think we should literally just have this in the description for the soft TTL.

```
| TTL Type |       Annotation Key       |         Annotation Value         |                                                       Description                                                      |
| -------- | -------------------------- | -------------------------------- | ---------------------------------------------------------------------------------------------------------------------- |
| Hard     | `ci.openshift.io/ttl.hard` | `time.Duration`-formatted string | The time after namespace creation when the namespace should be reaped.                                                 |
| Soft     | `ci.openshift.io/ttl.soft` | `time.Duration`-formatted string | The time after `Pod` termination or last active marker where, if no other `Pod`s are `Pending` or `Running`, the namespace should be reaped. The last active marker is optional and stored as a RFC3339-formatted date in the `ci.openshift.io/active` annotation. |
```

**Comment by petr-muller** [comment]
> @petr-muller do you want to review?

No need.

File: README.md [hunk 9]
```diff
@@ -14,6 +14,9 @@ Whichever TTL is reached first will be used by the controller to delete the name
 The controller will will only consider pods matching the `!ci.openshift.io/ttl.ignore` selector for soft TTL calculations, allowing
 non-matching pods to keep running while considering the namespace to be inactive.
 
+Inactive namespaces are being marked for deletion from the last pod's transition time, plus the softTTL time duration.
+However, if the `ci.openshift.io/active` annotation exists and it's after the latest pod's transition time, the controller will use that value for the calculations instead.
```
Comment (bbguimaraes) [inline README.md]: I agree that explaining the soft TTL again is a bit redundant. This is a good place to document the problem we are solving by introducing this annotation.

**Review by bbguimaraes — CHANGES_REQUESTED** [review]

**Comment by petr-muller** [comment]
My $0.02 on the description: I find Steve's suggestion from [this](https://github.com/openshift/ci-ns-ttl-controller/pull/2#discussion_r224445419) comment clearer than the current wording - it clearly describes how the soft ttl works with the new annotation and avoids complex (length) and vague ("recent") language.

So personally, I would use what Steve proposed + perhaps add a line somewhere about why the controller adds the annotation.

**Comment by droslean** [comment]
+1 @petr-muller, but the controller is not adding that annotation. It just uses its value is after the last transition time of the latest pod.

**Comment by droslean** [comment]
@petr-muller thanks, I missed @stevekuznetsov [comment](https://github.com/openshift/ci-ns-ttl-controller/pull/2#discussion_r224445419)

**Comment by petr-muller** [comment]
/lgtm
@stevekuznetsov needs to approve this anyway

> but the controller is not adding that annotation. It just uses its value is after the last transition time of the latest pod.

Yeah, my bad. Of course, on this side we do not need to care what puts the annotation there. I was not thinking decoupled enough.

**Pull request merged**